### ISSUE TITLE
MapFileWriter: OOM when processing ways

### ISSUE THREAD

**Issue by ludwigb** [mapsforge/mapsforge#920]
On our server we see the following OOM condition quite frequently:
```

Feb 08, 2017 2:17:01 AM org.mapsforge.map.writer.MapFileWriter processTile
WARNUNG: error in parallel preprocessing of ways
java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: Java heap space
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at org.mapsforge.map.writer.MapFileWriter.processTile(MapFileWriter.java:816)
	at org.mapsforge.map.writer.MapFileWriter.writeSubfile(MapFileWriter.java:906)
	at org.mapsforge.map.writer.MapFileWriter.writeFile(MapFileWriter.java:361)
	at org.mapsforge.map.writer.osmosis.MapFileWriterTask.complete(MapFileWriterTas
k.java:102)
	at org.openstreetmap.osmosis.set.v0_6.EntityMerger.run(EntityMerger.java:243)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.OutOfMemoryError: Java heap space
	at org.mapsforge.map.writer.util.PolyLabel.get(PolyLabel.java:124)
	at org.mapsforge.map.writer.MapFileWriter$WayPreprocessingCallable.call(MapFile
Writer.java:193)
	at org.mapsforge.map.writer.MapFileWriter$WayPreprocessingCallable.call(MapFile
Writer.java:94)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:11
42)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:6
17)
	... 1 more

Feb 08, 2017 2:17:40 AM org.mapsforge.map.writer.MapFileWriter$WayPreprocessingCallable
 call
INFORMATION: Large geometry 32951624755 (2835 coords, down from 3618 coords)
Feb 08, 2017 2:18:07 AM org.mapsforge.map.writer.MapFileWriter$WayPreprocessingCallable
 call

```

A quick look at the code shows that we do not handle the error correctly, as the way causing the OOM is silently ignored:

  ```
                  try {
                        List<Future<WayPreprocessingResult>> futures = EXECUTOR_SERVICE.invokeAll(callables);
                        for (Future<WayPreprocessingResult> wprFuture : futures) {
                            WayPreprocessingResult wpr;
                            try {
                                wpr = wprFuture.get();
                            } catch (ExecutionException e) {
                                LOGGER.log(Level.WARNING, "error in parallel preprocessing of ways", e);
                                continue;
                            }
                            if (wpr != null) {
                                wayBuffer.clear();
                                // increment count of ways on this zoom level                                             
                                entitiesPerZoomLevel[indexEntitiesPerZoomLevelTable][1]++;
                                if (configuration.isDebugStrings()) {
                                    writeWaySignature(wpr.getWay(), wayDataBuffer);
                                }
                                processWay(wpr, wpr.getWay(), currentTileLat, currentTileLon, wayBuffer);
                                // write size of way to way data buffer                                                   
                                wayDataBuffer.put(Serializer.getVariableByteUnsigned(wayBuffer.position()));
                                // write way data to way data buffer                                                      
                                wayDataBuffer.put(wayBuffer.array(), 0, wayBuffer.position());
                            }
                        }
                    } catch (InterruptedException e) {
                        LOGGER.log(Level.WARNING, "error in parallel preprocessing of ways", e);
                    }
```

The underlying reason for the error seems to be that we allocate as many threads as they are processors, on the server machine that might be 8. This then makes the thread run out of space.

I think we should terminate the process if this error condition happens, because correctness should never be compromised for efficiency. 

I also suggest that we reduce the thread pool size to 1. As I think the way processing is CPU bound it will probably not make much of a difference anyway.

**Comment by hgoebl** [comment]
👍 for reducing the number of threads. Let's hope this solves the problem for now.

What seams strange to me: OOM usually should terminate the JVM. Yes, we silently ignore errors, but a thing like OOM is an error which is caught only by `catch (Throwable ...)` or `Error`. So there must be another issue we should have an eye on. Is there a handler for uncaught exceptions somewhere which catches those OOMs?

I'd be interested in some statistics about the number of ways and POIs for the tiles. I know you said that logging is overwhelming already, but could you add a line per tile-write showing the number of ways and POIs, maybe with the total number of nodes and its coordinate? I have the feeling that we might have some "hot-spot" tiles which are crossed by many very long ways. If you think this is non-sense, then please just ignore it.

Still the question: why is germany.map built successfully and nordrhein-westfalen.map not? germany.map should face the same problems as its smaller part, right? Or is it a matter of a random condition, where doing 8 "hot-spot" tiles in parallel consumes all memory and in another run there is more luck by having the big tiles spread across many "runs" of the executor?

**Comment by ludwigb** [comment]
I think the java concurrent framework raises the OOM exception as a java.util.concurrent.ExecutionException, see logs above. 

> add a line per tile-write showing the number of ways and POIs, maybe with the total number of nodes and its coordinate?
Just considering how many tiles there are, this will generate a lot of output. The hot-spot tiles are very probable, particularly in the densely populated areas of Nordrhein Westfalen. I am just not sure if knowledge of this will help us further.

I was a bit surprised that the error producing the Nordrhein Westfalen map occurred again. I assume this might in some ways be an ordering problem, so by chance we have a number of long ways right after another (so they will cause the threads running in parallel at that time to overload), while in processing the larger area this situation is different. 

Let's see first if we still get errors. 

Unfortunately, the way the scripts are organized, it is easy to skip the map building  process when new files are already there, but pbf creation for larger sub-areas must still take place as we do not know if there are any smaller areas that need maps being built.

**Comment by devemux86** [comment]
@ludwigb do you mean in [MapFileWriter:353](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map-writer/src/main/java/org/mapsforge/map/writer/MapFileWriter.java#L353)?

The concurrency level is set to `availableProcessors * 2`, which in a 4-core with hyper-threading will return `8 * 2 = 16`! And if not set it defaults to 4.

**Comment by hgoebl** [comment]
I suppose it's [MapFileWriter:318](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map-writer/src/main/java/org/mapsforge/map/writer/MapFileWriter.java#L318)

**Comment by ludwigb** [comment]
Yes, the second. 
I have now for the sake of correctness set it to 1 and also kill the process if OOM conditions occur. 
Let's see what happens.

**Comment by devemux86** [comment]
Making some tests we could have a performance hit of around 50% when using 1 thread (depending the host).
So I introduced a new map writer `threads` parameter (default 1) for those who wish to use multithreading when building their maps.

Is there something additional to do regarding the exceptions?

**Comment by ludwigb** [comment]
Thanks,

actually I think terminating the process is correct, there should not be unhandled exceptions in the worker threads.

**Comment by ludwigb** [comment]
I will keep running my version for this run, I have not pushed the small changes that I made (setting threads to 1 and terminating process)

**Comment by devemux86** [comment]
> actually I think terminating the process is correct, there should not be unhandled exceptions in the worker threads.

So adding a `System.exit(1);` in [MapFileWriter:836](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map-writer/src/main/java/org/mapsforge/map/writer/MapFileWriter.java#L836)?

**Comment by ludwigb** [comment]
No, actually just rethrowing the exception as a RuntimeException, I think that will kill the process a bit more cleanly.

**Comment by devemux86** [comment]
I pushed the process termination too.

**Comment by devemux86** [comment]
Since all mentioned map-writer changes were implemented I'm closing here.
Please report any new findings.
